atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/expectations/KeyValueServiceDataTracker.java

```diff
@@ -0,0 +1,128 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.impl.expectations;
+
+import com.google.common.collect.Comparators;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+import com.google.common.util.concurrent.AtomicLongMap;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.transaction.api.expectations.ImmutableKvsCallReadInfo;
+import com.palantir.atlasdb.transaction.api.expectations.ImmutableTransactionReadInfo;
+import com.palantir.atlasdb.transaction.api.expectations.KvsCallReadInfo;
+import com.palantir.atlasdb.transaction.api.expectations.TransactionReadInfo;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Function;
+
+public final class KeyValueServiceDataTracker {
+    private final AtomicLongMap<TableReference> bytesReadByTable = AtomicLongMap.create();
+    private final LongAdder bytesReadOverall = new LongAdder();
+    private final AtomicLongMap<TableReference> kvsCallByTable = AtomicLongMap.create();
+    private final LongAdder kvsCallsOverall = new LongAdder();
+    private final ConcurrentMap<TableReference, KvsCallReadInfo> maximumBytesKvsCallInfoByTable =
+            new ConcurrentHashMap<>();
+    private final AtomicReference<Optional<KvsCallReadInfo>> maximumBytesKvsCallInfoOverall =
+            new AtomicReference<>(Optional.empty());
+
+    public TransactionReadInfo getReadInfo() {
+        return ImmutableTransactionReadInfo.builder()
+                .bytesRead(bytesReadOverall.longValue())
+                .kvsCalls(kvsCallsOverall.longValue())
+                .maximumBytesKvsCallInfo(maximumBytesKvsCallInfoOverall.get())
+                .build();
+    }
+
+    /**
+     * This is un-synchronized as it is called after task completion and the abort/commit stage.
+     * For returned statistics, no consistency guarantees are provided if the user interacts with the transaction
+     * outside the context of {@link com.palantir.atlasdb.transaction.api.TransactionTask} or after commit/abort
+     * stages (e.g. this can happen with futures/iterators).
+     */
+    public ImmutableMap<TableReference, TransactionReadInfo> getReadInfoByTable() {
+        Set<TableReference> tableRefs = Sets.intersection(
```

Comment:
Discussed on teams -- let's remove this, as `bytesReadByTable` & `kvsCallByTable` will return 0 instead of null so we do not need to worry about NPE. In addition, the intersection doesn't provide us with much benefit, outside of not reporting consistent results.